compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/TypesFactory.java

```diff
@@ -129,6 +134,27 @@ public TypeSymbol getTypeDescriptor(BType bType, boolean rawTypeOnly) {
                                                         bType.tsymbol.getName().getValue());
         }
 
+        if (this.typeCache.containsKey(bType)) {
+            TypeSymbol typeSymbol = this.typeCache.get(bType);
+
+            // Have to special case invokable types since equals() is overridden in BInvokableType.
+            if (bType.tag != TypeTags.INVOKABLE) {
+                return typeSymbol;
+            }
+
+            if (bType == ((AbstractTypeSymbol) typeSymbol).getBType()) {
+                return typeSymbol;
+            }
+        }
+
+        TypeSymbol typeSymbol = createTypeDescriptor(bType, moduleID);
+
+        // Because of the above explained reason, equivalent invokable types won't get cached either here.
+        typeCache.putIfAbsent(bType, typeSymbol);
```

Comment:
Does this means for the invokable types we just create the typeSymbol on query and does not replace the symbol? Will there be any harm for replacing the typeSymbol then?

Reply:
Since others don't override equals, this means we get type symbols cached for each unique `BType` instance. For invokable types, the `equals()` is basically a type equivalence test. For example, consider the following functions:
```ballerina
function add(int x, int y) returns int => x + y;
For the above  
function sum(int a, int b) returns int => a + b;
```
The above are two distinct functions. But the types of these are equivalent. Hence `equals()` will be true. So can't really cache the function type symbols against their internal counterparts. We can replace the cache entry. But still, for a set of equivalent invokable types, there will always be only one cache entry.

Reply:
Cool, then +1 for this approach